huaweicloud/services/apig/data_source_huaweicloud_apig_environments.go

```diff
@@ -0,0 +1,116 @@
+package apig
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/chnsz/golangsdk/openstack/apigw/dedicated/v2/environments"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/helper/hashcode"
+)
+
+func DataSourceEnvironments() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceEnvironmentsRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"instance_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z]([\w0-9]+)?$`),
```

Comment:
how about replace `([\w0-9]+)?` with `w?`, as `w` can match letters, digits and underscores (_)

Reply:
sure, the right expression is `^[A-Za-z](\w+)?$`.